### ISSUE TITLE
Replace tree control

### ISSUE THREAD

**Issue by tractorcow** [silverstripe/silverstripe-cms#744]
The current tree implementation is dated and could do with being replaced. Most likely, it would make more sense to shift to a react-based tree manager than relying on an upgrade of JSTree.

----

Originally: SiteTree Drag/Drop sorting usability (and bugs)

There are a few ongoing issues with the current drag-drop SiteTree panel.
- Dragging and dropping sometimes doesn't respond - Not sure why, but about half the time nothing happens (no ajax response / request). This is a seriously annoying bug.
- Dragging a page between two existing pages, and especially to the start and/or end of the sitetree is near impossible. My suggestion is that there should always be a drop "target" that  grows to fill the currently selected drop location, so that the result of releasing the mouse button can always be predicted in advance. There should be a drop target for each "nearby" possible drop location, so that users can select the one they intend to use.
- When attempting to place a page after a page with children this often expands the page's children instead. Can there be different drop locations for "insert after" and "expand children"?

This issue seriously aggravates my carpal tunnel.

**Comment by purplespider** [comment]
Any chance in getting this fixed for 3.2? It would be great not to have to teach users the "technique" in order to make drag & drop work! :)

**Comment by tractorcow** [comment]
I'll see if we can look at it. Time allowing. :)

**Comment by jonom** [comment]
I looked at updating this plugin once myself but it was too much for me - but I would be happy to help with any css/theming required if someone else can get the functionality up and running?

There are a few related issues that could all be addressed as part of a JSTree update:

* Missing TreeNode checkbox after reordering https://github.com/silverstripe/silverstripe-framework/issues/3524
* Tree sidebar right click stops working https://github.com/silverstripe/silverstripe-framework/issues/1545 (not sure if this is still an issue)
* The new page icon "status dots", don't update after a Multi-selection action https://github.com/silverstripe/silverstripe-cms/issues/1226 (don't think issue is caused by JSTree but it's related)
* Bulk action checkboxes are easy to mis-click https://github.com/silverstripe/silverstripe-framework/issues/4342

**Comment by jonom** [comment]
Holy dang, there are more than I though. Lots are related to the site tree not being refreshed properly after changes. Think it would make sense to update JSTree first, then fix as many of these issues as possible in one big update. Might require a little re-think of Hierarchy and trees in general in the framework though.

- https://github.com/silverstripe/silverstripe-cms/issues/1161
- https://github.com/silverstripe/silverstripe-cms/issues/1084
- https://github.com/silverstripe/silverstripe-cms/issues/1053
- https://github.com/silverstripe/silverstripe-cms/issues/867
- https://github.com/silverstripe/silverstripe-cms/issues/776
- https://github.com/silverstripe/silverstripe-cms/issues/767
- https://github.com/silverstripe/silverstripe-cms/issues/766
- https://github.com/silverstripe/silverstripe-cms/issues/763
- https://github.com/silverstripe/silverstripe-cms/issues/657
- https://github.com/silverstripe/silverstripe-cms/issues/572
- https://github.com/silverstripe/silverstripe-cms/issues/515
- https://github.com/silverstripe/silverstripe-cms/issues/508
- https://github.com/silverstripe/silverstripe-cms/issues/400

**Comment by tractorcow** [comment]
Wow there are a lot of issues... I wonder how much longer we can use this version of treeview before we have to upgrade. :D

**Comment by jonom** [comment]
Yeah I think because of the number of issues it might need a bit of a 'back to the drawing board' approach. The latest version of JSTree seems to have a pretty different API to the legacy (pre-release) version we're currently using, so if we're talking about updating to that it's probably worth doing some research to see if there is an alternative js-tree type plugin with more of the filtering functions etc. we need included out of the box - since either one will probably require a fair bit of work to drop in. Probably counter-productive to try and fix these issues *without* updating JSTree or switching to a different plugin as the fixes might become redundant and lead to re-work if/when the tree plugin is updated later.

**Comment by flashbackzoo** [comment]
Improving drag-and-drop is in our internal backlog and marked for 3.3. So will get some love over the next couple of months. And yeah - we'll have to consider fixing vs upgrading.

I ran into a similar fixing vs upgrading issue the other day with Chosen.js and dropdown fields. This and the Chosen issue are actually part of a wider issue, font-end dependency management, which currently doesn't exist. I'm keen to put together an RFC, using npm or bower, for 4.0. I think that would go some way to helping improving the upgrade path for the various libraries we use.

**Comment by jonom** [comment]
Noice! Will look forward to that @flashbackzoo

**Comment by flashbackzoo** [comment]
@jonom https://github.com/silverstripe/silverstripe-framework/issues/4372

**Comment by jonom** [comment]
Cheers @flashbackzoo I don't have the relevant knowledge to comment on the specifics but I see lots of discussion going on so that's great ;)

**Comment by colymba** [comment]
Just a wild idea... do we really need jsTree? I mean, can we not just ditch it and get away with just a drag-n-drop library? Do we need more than simple drag-n-drop js callbacks and all the rest can be done with simple CSS?

There are a bunch or library that we could look into like:
http://johnny.github.io/jquery-sortable/
http://lukasoppermann.github.io/html5sortable/examples/index.html
http://rubaxa.github.io/Sortable/

**Comment by tractorcow** [comment]
If someone wants to rewrite all of the js menu code in CMS then sure. :P We also have right click menu integration, bulk selection, partial tree ajax loading, etc... so those would need to be re-implemented.

**Comment by clarkepaul** [comment]
Some good alternatives there.

**Comment by tractorcow** [comment]
Todo : Replace drag drop tree with new library. :)

**Comment by kinglozzer** [comment]
Just a heads up, JSTree 3.x requires jQuery 1.9 or newer and we’re still using 1.7.2 in SS4. It might be a good excuse to update jQuery, even if it’s just a few versions newer rather than a “current” release, and move it into our build process properly

**Comment by dhensby** [comment]
> It might be a good excuse to update jQuery

This has been discussed, the biggest challenge is getting entwine to work with newer versions of jQuery see #3384

I think the overall feeling of the core team is that, as we move away from jQuery towards React, spending development effort on upgrading JSTree, Entwine and jQuery is not productive. The goal is to remove these components rather than support them.

**Comment by tractorcow** [comment]
I'd rather make a react-sitetree manager to be honest, than invest more in react.

**Comment by chillu** [comment]
FYI I'm advocating to remove the ability for a tree to show search results in SS4 - we've already simplified TreeDropdownField into a list-based UI (although it still uses `Hierarchy` and tree marking behind the scenes). Next cab off the rank is removal of tree search results when editing a page, and when showing full screen search results: https://github.com/silverstripe/silverstripe-cms/issues/1926

**Comment by NightJar** [comment]
Has this been done in `4.3.0`, or `4` in general? @chillu

**Comment by chillu** [comment]
I think it's been released with 4.2.0?

**Comment by NightJar** [comment]
Perhaps, unsure. Either way this is effectively done and could be closed though right? :)

**Comment by ScopeyNZ** [comment]
It might still be pertinent to keep an issue open for the eventual conversion of the sitetree into React. I'll leave that to @chillu's judgement though.

**Comment by chillu** [comment]
Actually, yeah let's keep it open. I only read the last few comments in the thread, and was under the impression that it was limited to the treedropdownfield tree view, which we've fixed. But the scope is all tree implementations

**Comment by michalkleiner** [comment]
Just a suggestion I came across the other day, at least as an inspiration: frontend-collective/react-sortable-tree

**Comment by chillu** [comment]
Thanks Michal! I think the main bit missing from that UI library would be lazy loading of partial trees. Which has been [discussed](https://github.com/frontend-collective/react-sortable-tree/issues/211) in 2017 on that repo. Also, 500 issues in its lifetime. Open source is hard :)

**Comment by GuySartorelli** [comment]
This is an ancient issue about something we have no immediate plans to do anything about.
At some stage we'll need to either replace or upgrade jstree - but when we do so I think we'll just make a new issue. A lot of this issue is talking about how there's lots of bugs, but almost all of the linked issues are resolved, so the discussion here probably isn't useful anymore.

Closing.
